pandas/core/reshape/pivot.py

```
@@ -534,67 +551,36 @@ def _normalize(table, normalize, margins, margins_name='All'):
+    }
 
-        table = f(table)
-        table = table.fillna(0)
+    if margins is False:
+        normalizers['all'] = lambda x: x / x.sum(axis=1).sum(axis=0)
 
     elif margins is True:
+        # skip margin rows and cols for normalization
+        normalizers['all'] = lambda x: x / x.iloc[:-1, :-1].sum(axis=1)\
+                                                           .sum(axis=0)
 
-        column_margin = table.loc[:, margins_name].drop(margins_name)
-        index_margin = table.loc[margins_name, :].drop(margins_name)
-        table = table.drop(margins_name, axis=1).drop(margins_name)
-        # to keep index and columns names
-        table_index_names = table.index.names
-        table_columns_names = table.columns.names
-
-        # Normalize core
-        table = _normalize(table, normalize=normalize, margins=False)
-
-        # Fix Margins
-        if normalize == 'columns':
-            column_margin = column_margin / column_margin.sum()
-            table = concat([table, column_margin], axis=1)
-            table = table.fillna(0)
-
-        elif normalize == 'index':
-            index_margin = index_margin / index_margin.sum()
-            table = table.append(index_margin)
-            table = table.fillna(0)
-
-        elif normalize == "all" or normalize is True:
-            column_margin = column_margin / column_margin.sum()
-            index_margin = index_margin / index_margin.sum()
-            index_margin.loc[margins_name] = 1
-            table = concat([table, column_margin], axis=1)
-            table = table.append(index_margin)
-
-            table = table.fillna(0)
+    else:
+        raise ValueError("Not a valid margins argument: {!r}".format(margins))
 
-        else:
-            raise ValueError("Not a valid normalize argument")
+    normalizers[True] = normalizers['all']
 
-        table.index.names = table_index_names
-        table.columns.names = table_columns_names
+    try:
+        f = normalizers[normalize]
```

this is an internal error yes? (IOW is NOT exposed to the user). is there a test? 
I would have just raise KeyError if it fails (IOW its not found)